File: libraries/Willow.Batch/src/Dtos/BatchRequestDto.cs
namespace Willow.Batch
{
    using System;

    /// <summary>
    /// A batch request.
    /// </summary>
    public class BatchRequestDto
    {
        /// <summary>
        /// Gets the specifications on how to sort the batch.
        /// </summary>
        public SortSpecificationDto[] SortSpecifications { get; init; } = Array.Empty<SortSpecificationDto>();

        /// <summary>
        /// Gets or sets the specification on how to filter the batch.
        /// </summary>
        public FilterSpecificationDto[] FilterSpecifications { get; set; } = Array.Empty<FilterSpecificationDto>();

        /// <summary>
        /// Gets the page number to return for the batch (one-based).
        /// </summary>
        public int? Page { get; init; }

        /// <summary>
        /// Gets the amount of items in the batch.
        /// </summary>
        public int? PageSize { get; init; }
    }
}

File: applications/UserManagement/Authorization.TwinPlatform.Web/authorization-client-app/src/types/BatchRequestDto.ts
import { GridFilterModel, GridSortModel } from "@willowinc/ui";

export class BatchRequestDto {
  sortSpecifications: SortSpecificationDto[];
  filterSpecifications: FilterSpecificationDto[];
  page: number;
  pageSize: number;
  constructor(); // no arguments
  constructor(filterSpecifications?: FilterSpecificationDto[], sortSpecifications?: SortSpecificationDto[], page?: number, pageSize?: number);

  constructor(filterSpecifications?: FilterSpecificationDto[], sortSpecifications?: SortSpecificationDto[], page?: number, pageSize?: number) {
    this.sortSpecifications = sortSpecifications ?? [];
    this.filterSpecifications = filterSpecifications ?? [];
    this.page = page ?? 0;
    this.pageSize = pageSize ?? 100;
  }
}


export class SortSpecificationDto {
  field: string;
  sort?: string | undefined | null;

  constructor(field: string, sort?: string| undefined | null) {
    this.field = field;
    this.sort = sort;
  }

  static MapFrom(gridSortModel: GridSortModel) {
    return gridSortModel.map(m => new SortSpecificationDto(m.field, m.sort));
  }
}

export class FilterSpecificationDto {
  operator!: string;
  field: string;
  logicalOperator?: string | undefined;
  value?: any | undefined;
  isQuickFilter: boolean;

  constructor(field: string, operator: string, value: any, logicalOperator: string | undefined, isQuickFilter?:boolean) {
    this.field = field;
    this.operator = FilterSpecificationDto.ConvertToKnowOperator(operator);
    this.value = value ?? '';
    this.logicalOperator = logicalOperator;
    this.isQuickFilter = isQuickFilter ?? false;
  }

  static MapFrom(filterPropertyModel: GridFilterModel) {
    return filterPropertyModel.items.map(m => new FilterSpecificationDto(m.field, m.operator, m.value, filterPropertyModel.logicOperator));
  }

  static ConvertToKnowOperator(op: string) {
    if (op === 'isAnyOf')
      return 'in';
    else
      return op;
  }

}
